src/persist-client/src/impl/encoding.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::trace::Description;
+use mz_persist_types::{Codec, Codec64};
+use mz_proto::{IntoRustIfSome, ProtoType, RustType, TryFromProtoError};
+use prost::Message;
+use serde::{Deserialize, Serialize};
+use timely::progress::{Antichain, Timestamp};
+use timely::PartialOrder;
+use uuid::Uuid;
+
+use crate::error::CodecMismatch;
+use crate::r#impl::state::{
+    HollowBatch, ProtoHollowBatch, ProtoHollowBatchPart, ProtoReader, ProtoSnapshotSplit,
+    ProtoStateRollup, ProtoTrace, ProtoU64Antichain, ProtoU64Description, ReadCapability, State,
+    StateCollections,
+};
+use crate::r#impl::trace::Trace;
+use crate::read::{ReaderId, SnapshotSplit};
+use crate::ShardId;
+
+pub(crate) fn parse_id(id_prefix: char, id_type: &str, encoded: &str) -> Result<[u8; 16], String> {
+    let uuid_encoded = match encoded.strip_prefix(id_prefix) {
+        Some(x) => x,
+        None => return Err(format!("invalid {} {}: incorrect prefix", id_type, encoded)),
+    };
+    let uuid = Uuid::parse_str(&uuid_encoded)
+        .map_err(|err| format!("invalid {} {}: {}", id_type, encoded, err))?;
+    Ok(*uuid.as_bytes())
+}
+
+impl RustType<String> for ShardId {
+    fn into_proto(&self) -> String {
+        self.to_string()
+    }
+
+    fn from_proto(proto: String) -> Result<Self, TryFromProtoError> {
```

Comment:
okay now that I'm staring at this again, I think I don't understand why we have methods named `from_proto` that don't pass in a proto as a parameter

Reply:
fwiw that's the mz convention, not anything I came up with :)

it seems to be because it's the "proto representation" of the thing, which might or might not be a proto. e.g. in mz_proto there's an impl to map "rust usize" <-> "proto u64"